pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+		jm.queue.AddRateLimited(key)
+	case requeueAfter != nil:
+		jm.queue.Forget(key)
+		jm.queue.AddAfter(key, *requeueAfter)
+	}
+	return true
+}
+
+func (jm *ControllerV2) sync(cronJobKey string) (*time.Duration, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(cronJobKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cronJob, err := jm.cronJobLister.CronJobs(ns).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// may be cronjob is deleted, dont need to requeue this key
+		klog.V(4).InfoS("cronjob not found, may be it is deleted", "cronjob", klog.KRef(ns, name), "err", err)
+		return nil, nil
+	case err != nil:
+		// for other transient apiserver error requeue with exponential backoff
+		return nil, err
+	}
+
+	jobsToBeReconciled, err := jm.getJobsToBeReconciled(cronJob)
+	if err != nil {
+		return nil, err
+	}
+
+	cronJob, requeueAfter, err := jm.syncCronJob(cronJob, jobsToBeReconciled)
+	if err != nil {
+		klog.V(2).InfoS("error reconciling cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "err", err)
+		return nil, err
+	}
+
+	err = jm.cleanupFinishedJobs2(cronJob, jobsToBeReconciled)
+	if err != nil {
+		klog.V(2).InfoS("error cleaning up jobs", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "resourceVersion", cronJob.GetResourceVersion(), "err", err)
+		return nil, err
+	}
+
+	if requeueAfter != nil {
+		klog.V(4).InfoS("re-queuing cronjob", "cronjob", klog.KRef(cronJob.GetNamespace(), cronJob.GetName()), "requeueAfter", requeueAfter)
+		return requeueAfter, nil
+	}
+	// this marks the key done, currently only happens when the cronjob is suspended or spec has invalid schedule format
+	return nil, nil
+}
+
+// resolveControllerRef returns the controller referenced by a ControllerRef,
+// or nil if the ControllerRef could not be resolved to a matching controller
+// of the correct Kind.
+func (jm *ControllerV2) resolveControllerRef(namespace string, controllerRef *metav1.OwnerReference) *batchv1beta1.CronJob {
+	// We can't look up by UID, so look up by Name and then verify UID.
+	// Don't even try to look up by Name if it's the wrong Kind.
+	if controllerRef.Kind != controllerKind.Kind {
+		return nil
+	}
+	cronJob, err := jm.cronJobLister.CronJobs(namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil
+	}
+	if cronJob.UID != controllerRef.UID {
+		// The controller we found with this Name is not the same one that the
+		// ControllerRef points to.
+		return nil
+	}
+	return cronJob
+}
+
+func (jm *ControllerV2) getJobsToBeReconciled(cronJob *batchv1beta1.CronJob) ([]batchv1.Job, error) {
+	var jobSelector labels.Selector
+	if len(cronJob.Spec.JobTemplate.Labels) == 0 {
+		jobSelector = labels.Everything()
+	} else {
+		jobSelector = labels.Set(cronJob.Spec.JobTemplate.Labels).AsSelector()
+	}
+	jobList, err := jm.jobLister.Jobs(cronJob.Namespace).List(jobSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	jobsToBeReconciled := []batchv1.Job{}
```

Comment:
The lister standard users `[]*` to avoid copies.

Reply:
do you mind if I take this as a follow on? 

This list is used by code in the old controller which would need refactoring as well or copy-pasting

Reply:
addressed in the follow-on PR